wave_lang/kernel/wave/gather_to_shared.py

```diff
@@ -438,3 +448,98 @@ def gather_to_shared(
         update_write_dependencies(new_writes, trace)
 
     DCE(trace)
+
+
+def gather_to_shared_swizzling(
+    trace: CapturedTrace,
+    constraints: list[Constraint],
+    options: WaveCompileOptions,
+):
+    """
+    This pass is used to swizzle the gather to shared op global index and
+    corresponding LDS load index to reduce LDS bank conflicts.
+
+    The formula for swizzling is:
+    ```
+    new_col = xor(row % max_phase, col // elements_per_thread) * elements_per_thread
+    ```
+    """
+    if "gfx95" not in options.target:
+        logger.info("gather_to_shared_swizzling not supported on this architecture")
+        return
+
+    logger.info("gather_to_shared_swizzling")
+
+    id_to_gather = defaultdict(list)
+    for gather in trace.walk(lambda x: isinstance(get_custom(x), GatherToLDS)):
+        gather = get_custom(gather)
+        key = gather.pre_expansion_id
+        id_to_gather[key].append(gather)
+
+    if not id_to_gather:
+        return
+
+    for gathers in id_to_gather.values():
+        mem = gathers[0].dst
+        reads = [
+            get_custom(read) for read in mem.users if isinstance(get_custom(read), Read)
+        ]
+        gather = gathers[0]
+        read = reads[0]
+        if (
+            gather.dtype != read.dtype
+            or gather.elements_per_thread != read.elements_per_thread
+        ):
+            logger.info(
+                "mismatched gather and read thread shapes: "
+                f"gather.dtype={gather.dtype}, read.dtype={read.dtype}, "
+                f"gather.elements_per_thread={gather.elements_per_thread}, "
+                f"read.elements_per_thread={read.elements_per_thread}"
+            )
+            continue
+
+        elements_per_thread = gather.elements_per_thread
+        logger.info(f"elements_per_thread={elements_per_thread}")
+
+        if elements_per_thread * gather.dtype.bitwidth() != 128:
+            logger.info(
+                f"gather to shared swizzling only supported for 128-bit elements, got {elements_per_thread}x{gather.dtype}"
+            )
+            continue
+
+        shape = get_custom(mem).type.symbolic_shape
+        if len(shape) < 2:
+            logger.info(f"shape={shape} must be at least 2D")
+            continue
+
+        col_dim = infer_dim(shape[-1])
+        row_dim = infer_dim(shape[-2])
+
+        max_phase = 8
+
+        for read in reads:
+            index = remove_global_indexing(read.index, constraints)
```

Comment:
NIT/thought exercise: Thoughts on moving this pass after `apply_shared_memory_indexing_corrections` so we may not need to remove the global indexing?

Reply:
I would prefer to just move `apply_shared_memory_indexing_corrections ` earlier (or just do it as part of promotion)